pkg/registry/pod/rest.go

```diff
@@ -81,9 +81,28 @@ func (podStrategy) ValidateUpdate(ctx api.Context, obj, old runtime.Object) fiel
 	return append(errorList, validation.ValidatePodUpdate(obj.(*api.Pod), old.(*api.Pod))...)
 }
 
-// CheckGracefulDelete allows a pod to be gracefully deleted.
+// CheckGracefulDelete allows a pod to be gracefully deleted. It updates the DeleteOptions to
+// reflect the desired grace value.
 func (podStrategy) CheckGracefulDelete(obj runtime.Object, options *api.DeleteOptions) bool {
-	return false
+	if options == nil {
+		return false
+	}
+	pod := obj.(*api.Pod)
+	// use the user provided value
+	if options.GracePeriodSeconds != nil {
+		period := int64(*options.GracePeriodSeconds)
+		// if the pod is not currently being deleted, or the value isn't set, or the user value is less than the stored value
+		if pod.Spec.TerminationGracePeriodSeconds == nil || pod.DeletionTimestamp == nil || period < *pod.Spec.TerminationGracePeriodSeconds {
+			pod.Spec.TerminationGracePeriodSeconds = &period
```

Comment:
> On May 28, 2015, at 3:56 PM, Brian Grant <notifications@github.com> wrote:
> 
> In pkg/registry/pod/rest.go:
> 
> >  func (podStrategy) CheckGracefulDelete(obj runtime.Object, options *api.DeleteOptions) bool {
> > -	return false
> > +	if options == nil {
> > +		return false
> > +	}
> > +	pod := obj.(*api.Pod)
> > +	// use the user provided value
> > +	if options.GracePeriodSeconds != nil {
> > +		period := int64(*options.GracePeriodSeconds)
> > +		// if the pod is not currently being deleted, or the value isn't set, or the user value is less than the stored value
> > +		if pod.Spec.TerminationGracePeriodSeconds == nil || pod.DeletionTimestamp == nil || period < *pod.Spec.TerminationGracePeriodSeconds {
> > +			pod.Spec.TerminationGracePeriodSeconds = &period
> We need 2 fields.
> 
> TerminationGracePeriodSeconds should be the starting default grace period. It can be updated normally, even increased. If updated once the grace period has started, the update has no effect.
> 
> The second field, which would be new, would be initialized from either TerminationGracePeriodSeconds or from DeleteOptions.GracePeriodSeconds upon deletion, and could only be decreased once set.
> 
> As in the node controller, Kubelet probably has to initialize its own local timer once it observes the 2nd field (or DeletionTimestamp) has been set.
> 
> Yes, it's another field, but so much simpler.
> 
Ok.  Do we want this in meta (my preference is yes), and do we want to call it DeletionGracePeriodSeconds?
> —
> Reply to this email directly or view it on GitHub.
>